Styleable::renderer does a linear search over ::view-transition-groups.

mattwoodrow opened the pull request:
#### b296bf89104e6eaa85c0be3adfc31f11f0b5d906
<pre>
Styleable::renderer does a linear search over ::view-transition-groups.
<a href="https://bugs.webkit.org/show_bug.cgi?id=282410">https://bugs.webkit.org/show_bug.cgi?id=282410</a>
&lt;<a href="https://rdar.apple.com/139031118">rdar://139031118</a>&gt;

Reviewed by Tim Nguyen.

This can be arbitrarily long, we should use a hash-based lookup instead.

Introduces a new RenderBlockFlow subclass, RenderViewTransitionRoot for the
::view-transition psuedo element.

Stores a hashtable of the ::view-transition-group children renderers attached to
the root, and makes Styleable::renderer() use this for faster lookup.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderViewTransitionRoot const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::viewTransitionRoot const):
(WebCore::RenderView::setViewTransitionRoot):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/RenderViewTransitionRoot.cpp: Added.
(WebCore::RenderViewTransitionRoot::RenderViewTransitionRoot):
(WebCore::RenderViewTransitionRoot::addChildGroup):
(WebCore::RenderViewTransitionRoot::removeChildGroup):
(WebCore::RenderViewTransitionRoot::childGroupForName):
* Source/WebCore/rendering/RenderViewTransitionRoot.h: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::renderer const):

Canonical link: <a href="https://commits.webkit.org/286166@main">https://commits.webkit.org/286166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ded7286085105292503c36b2488f9a0b199b848

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74996 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54426 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79433 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26235 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77113 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63562 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2211 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58882 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17143 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78063 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49031 "Found 1 new test failure: media/media-garbage-collection.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64426 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39268 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21927 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24567 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67476 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80914 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2314 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1402 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67134 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64445 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66435 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16518 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10371 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8532 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2279 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2307 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3228 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -147,15 +148,7 @@ RenderElement* Styleable::renderer() const
-        }
-
-        // If we can't find the correct group, return nullptr.
+        RenderBlockFlow* correctGroup = viewTransitionRoot->childGroupForName(pseudoElementIdentifier->nameArgument);
```
```suggestion
        auto* correctGroup = viewTransitionRoot->childGroupForName(pseudoElementIdentifier->nameArgument);
```
(or CheckedPtr)

mattwoodrow committed:
Styleable::renderer does a linear search over ::view-transition-groups.
https://bugs.webkit.org/show_bug.cgi?id=282410
<rdar://139031118>

Reviewed by Tim Nguyen.

This can be arbitrarily long, we should use a hash-based lookup instead.

Introduces a new RenderBlockFlow subclass, RenderViewTransitionRoot for the
::view-transition psuedo element.

Stores a hashtable of the ::view-transition-group children renderers attached to
the root, and makes Styleable::renderer() use this for faster lookup.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderViewTransitionRoot const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::viewTransitionRoot const):
(WebCore::RenderView::setViewTransitionRoot):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/RenderViewTransitionRoot.cpp: Added.
(WebCore::RenderViewTransitionRoot::RenderViewTransitionRoot):
(WebCore::RenderViewTransitionRoot::addChildGroup):
(WebCore::RenderViewTransitionRoot::removeChildGroup):
(WebCore::RenderViewTransitionRoot::childGroupForName):
* Source/WebCore/rendering/RenderViewTransitionRoot.h: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::renderer const):

Canonical link: https://commits.webkit.org/286166@main
